ReSharper.FSharp/src/TypeProvidersLoader/Protocol/ModelCreators/ProvidedRdModelsCreatorWithCacheBase.cs

```diff
@@ -0,0 +1,43 @@
+﻿using JetBrains.Annotations;
+using JetBrains.ReSharper.Plugins.FSharp.TypeProvidersLoader.Protocol.Cache;
+
+namespace JetBrains.ReSharper.Plugins.FSharp.TypeProvidersLoader.Protocol.ModelCreators
+{
+  public abstract class ProvidedRdModelsCreatorWithCacheBase<T, Tu> : IProvidedRdModelsCreatorWithCache<T, Tu, int>
+    where Tu : class
+  {
+    private readonly IProvidedCache<(T, int), int> myCache;
+    private int myCurrentId;
+
+    protected ProvidedRdModelsCreatorWithCacheBase(IProvidedCache<(T, int), int> cache) =>
+      myCache = cache;
+
+    [ContractAnnotation("providedModel:null => null")]
+    public virtual Tu CreateRdModel(T providedModel, int typeProviderId)
+    {
+      if (providedModel == null) return null;
+
+      var id = CreateEntityKey(providedModel);
+      var model = CreateRdModelInternal(providedModel, id, typeProviderId);
+
+      myCache.Add(id, (providedModel, typeProviderId));
+
+      return model;
+    }
+
+    public virtual int GetOrCreateId(T providedModel, int typeProviderId)
+    {
+      if (providedModel == null) return default;
+
+      var id = CreateEntityKey(providedModel);
+
+      myCache.Add(id, (providedModel, typeProviderId));
+      return id;
+    }
+
+    //TODO: assert threading
```

Comment:
For the record: we've agreed that we don't want to assert it everywhere, but it would be cool to at least guard this increment. If easily achievable, of course.